### PR TITLE
BFB-527 & BFB-501 - Integrate Countdown Timer and Ring System

### PR DESCRIPTION
**Pull request by unoctium1**
**JIRA Ticket**
https://perigongames.atlassian.net/browse/BFB-501
https://perigongames.atlassian.net/browse/BFB-527

**Description**
Integrate rings with countdown timer. Each ring system now has a specified duration and will reset if not complete when the duration ends. Also add some functionality to countdown timer

### CODE REVIEW

File: Assets/_BForBoss/_Core/Scripts/HUD/CountdownViewBehaviour.cs [hunk 1]
```diff
@@ -12,19 +12,36 @@ public class CountdownViewBehaviour : MonoBehaviour
         private bool _isRunning;
 
         private float _amountOfTime;
+        
+        private Action OnCountdownCompleted;
+        
+        public bool IsRunning => _isRunning;
 
-        public void StartCountdown(float amountOfTime)
+        public void StartCountdown(float amountOfTime, Action onCountdownCompleted = null)
```
Comment (m4rtint) [inline Assets/_BForBoss/_Core/Scripts/HUD/CountdownViewBehaviour.cs:20]: Imo, the count down HUD / UI, shouldn't be the one that determines it ended. 

The logic of "The time has ended, now do this" should be contained within the Manager itself. 
This UI/HUD behaviour should soley be in charge of letting users know the time.

Comment (unoctium1) [reply Assets/_BForBoss/_Core/Scripts/HUD/CountdownViewBehaviour.cs:20]: Yeah that was mostly a result of how CountdownViewBehaviour currently is setup, I can rewrite it to be more of an actual view though

Comment (m4rtint) [reply Assets/_BForBoss/_Core/Scripts/HUD/CountdownViewBehaviour.cs:20]: I just remembered countDownViewBehaviour is more than not a placeholder. This part is flexible

File: Assets/_BForBoss/_Core/Scripts/Labor/RingSystem/RingLaborManager.cs [hunk 2]
```diff
@@ -29,22 +27,31 @@ public void Reset()
             _laborSystem = new LaborSystem(_ringSystems);
         }
         
-        public void Initialize()
+        public void Initialize(CountdownViewBehaviour countdownView)
```
Comment (m4rtint) [inline Assets/_BForBoss/_Core/Scripts/Labor/RingSystem/RingLaborManager.cs:30]: `CountdownViewBehaviour` 

Make this a singleton, or just call its static shared instance.

**Review by Gursaanj — APPROVED** [review]

File: Assets/_BForBoss/_Core/Scripts/Labor/RingSystem/RingLaborManager.cs (hunk 2 above)
Comment (unoctium1) [reply Assets/_BForBoss/_Core/Scripts/Labor/RingSystem/RingLaborManager.cs]: Split countdown view into a singleton countdown timer, and turned the view into more of a view for that

File: Assets/_BForBoss/_Core/Scripts/Labor/RingSystem/RingLaborManager.cs [hunk 3]
```diff
@@ -35,16 +32,24 @@ public void Initialize()
             _laborSystem = new LaborSystem(_ringSystems);
         }
 
+        public void ToggleTimer()
+        {
+            if(CountdownTimer.Instance.IsRunning)
```
Comment (m4rtint) [inline Assets/_BForBoss/_Core/Scripts/Labor/RingSystem/RingLaborManager.cs:37]: if you had to call this many `CountdownTimer.instance`, you can probably put this whole method intothe `countdownTimer`itself.

Comment (unoctium1) [reply Assets/_BForBoss/_Core/Scripts/Labor/RingSystem/RingLaborManager.cs]: Yeah good call, the toggle functionality will be a lot cleaner inside it as well

**Review by m4rtint — APPROVED** [review]

**Pull request merged**